netket/nn/symmetric_linear.py

```diff
@@ -101,62 +101,446 @@ def full_bias(self, bias):
+    irrep matrices. After Fourier transforming, the convolution translates to
+    :: math ::
+        y^{(i)}_\rho = \sum_j f^{(j)}_\rho W^{(ij)}_\rho,
+    where all terms are d x d matrices rather than numbers, and the juxtaposition
+    stands for matrix multiplication.
     """
 
-    symmetry_info: Union[HashableArray, PermutationGroup]
+    irreps: Tuple[HashableArray]
+    """Irrep matrices of the symmetry group. Each element of the list is an
+    array of shape [n_symm, d, d]; irreps[i][j] is the representation of the
+    jth group element in irrep #i."""
+    in_features: int
+    """The number of input features; must be the second dimension of the input."""
+    out_features: int
+    """The number of output features; returned as the second dimension of the output."""
+    use_bias: bool = True
+    """Whether to add a bias to the output (default: True)."""
+    mask: Optional[HashableArray] = None
+    """Mask that zeros out elements of the filter. Should be of shape [inp.shape[-1]]"""
+
+    dtype: DType = jnp.float64
+    """The dtype of the weights."""
+    precision: Any = None
+    """numerical precision of the computation see `jax.lax.Precision`for details."""
+
+    kernel_init: Optional[NNInitFunc] = None
+    """Initializer for the Dense layer matrix."""
+    bias_init: Optional[NNInitFunc] = None
+    """Initializer for the bias."""
+
+    def setup(self):
+        self.n_symm = self.irreps[0].shape[0]
+        self.forward = jnp.concatenate(
+            [jnp.asarray(irrep).reshape(self.n_symm, -1) for irrep in self.irreps],
+            axis=1,
+        )
+        self.inverse = jnp.concatenate(
+            [
+                jnp.asarray(irrep).conj().reshape(self.n_symm, -1)
+                * (irrep.shape[-1] / self.n_symm)
+                for irrep in self.irreps
+            ],
+            axis=1,
+        ).transpose()
+
+        # Convert between vectors of length n_symm and tuples of arrays of shape
+        # n_irrep × irrep_size^2
+        self.assemble = lambda arrays: jnp.concatenate(
+            [array.reshape(array.shape[:-3] + (-1,)) for array in arrays], axis=-1
+        )
+
+        irrep_size = 1
+        n_same_size = 0
+        shapes = []
+        for irrep in self.irreps:
+            if irrep_size == irrep.shape[-1]:
+                n_same_size += 1
+            else:
+                shapes.append((n_same_size, irrep_size, irrep_size))
+                irrep_size = irrep.shape[-1]
+                n_same_size = 1
+        shapes.append((n_same_size, irrep_size, irrep_size))
+        limits = np.cumsum([0] + [np.prod(shape) for shape in shapes])
+
+        self.disassemble = lambda vecs: tuple(
+            vecs[..., limits[i] : limits[i + 1]].reshape(vecs.shape[:-1] + shape)
+            for i, shape in enumerate(shapes)
+        )
+
+        self.stdev = 1.0 / np.sqrt(self.in_features * self.n_symm)
+
+    def forward_ft(self, inputs: Array, dtype: DType) -> Tuple[Array]:
```

Comment:
remove `dtype` from the call sequence for it isn't used anywhere? (and `jnp.tensordot` takes care of type promotion)